packages/core/src/Node.ts

```diff
@@ -141,11 +141,13 @@ export abstract class BaseNode<T extends Metadata = Metadata> {
   }
 
   /**
-   * Used with built in JSON.stringify
+   * Creates a deep-clone of the node as JSON
    * @returns
    */
   toJSON(): Record<string, any> {
-    return { ...this, type: this.getType() };
+    const json: Record<string, any> = structuredClone(this);
```

Comment:
@yisding I think the core issue is that this function did not return a deep clone.
Without this change, the test case `nodeToMetadata should not modify a node's metadata` would fail. 

Not using `structuredClone`, and instead doing an explicit deep clone in each node class would be better, but I think we need better unit testing for such a structural change.

Besides the unit tests, I also tested this PR with MongoDB vector store and `VectorStoreIndex`. 
As far as I know, the MongoDB load took longer than last time. it might be because of the deep cloning, but functionally, all tests passed.

If you're ok with these changes, then we can build a release.

Reply:
OK so `toJSON` is a special function that's specifically designed to be used with JSON.stringify.

https://javascript.info/json

So inherently for that purpose a deep clone wouldn't be necessary because the object is being turned into a string anyways.

However, we're calling toJSON directly in the other function to further modify before stringification. I think what we should do is move the structuredClone outside of this function and into the caller.

Reply:
@yisding, great catch! I encapsulated these serialization functions and documented them in BaseNode, so we're not running into similar issues again.